### ISSUE TITLE
Keycloak 19 (Quarkus) failing to load Admin Console after upgrade from v14

### ISSUE THREAD

**Issue by mortenjoenby** [keycloak/keycloak#13937]
### Describe the bug

We are upgrading from Keycloak 14 (Wildfly) w/ PostgreSQL 11 to Keycloak 19 (Quarkus) w/ PostgreSQL 14. 
We "migrated" the database by using pg_dump and pg_restore from the old to a new database instance.
Everything is running in AKS 1.22.6 and we're using a 19.0.1 image.

Important: we have 333 realms!

After a few attempts getting the upgrade to run through - by changing startupProbe failure threshold - we have a running Keycloak 19, BUT ... Admin Console is not loading, meaning that we get to see the initial screen, but realms are NEVER loading.
We can't see in the logs (container level) what Keycloak is doing (if anything), but we can't move on from here.

I posted in the keycloak-user list - "Keycloak 14 (Wildfly) to 19 (Quarkus) upgrade times out" - and got a couple of replies indicating that Keycloak 19 is still experiencing "issues", and I was recommended to stay with Keycloak 18.
Perhaps that's a fact, but I find it strange if we can't succeed in using v19 as it's officially released.

We have just now tried with a database with 70 realms. Going to v18 works fine, going to v19 also works fine.

We are about to try with yet another database with 296 realms.

ANY feedback to this is highly appreciated!

### Version

14, 19

### Expected behavior

We of course expect an upgrade to v19 to work smoothly, and that the Admin Console is able to load all realms and work fine.

### Actual behavior

Admin Console not loading realms when having "many" realms.

### How to Reproduce?

Not sure honestly, but we suspect the amount of realms to be the culprit.

**Comment by mortenjoenby** [comment]
Doing a bit more research - as we are now running on the 3rd hour of exporting via pg_dump - we have found that our EVENT_ENTITY table is basically the WHOLE database - 36 GB on one DB, 80GB on another.
This can perhaps explain our general bad performance on our different Keycloak instances with many realms.
We realize that we have enabled all events on all realms with NO expiration/retention!
We will of course change this, but is there a nice way to easily clear all these?
Could this also be why Admin Console is not loading any data after the upgrade?

**Comment by pedroigor** [comment]
@mortenjoenby You can try to disable the new administration console to compare the behavior when using the legacy one. For that, when starting the server you can use the `features-disabled=admin2` option.

**Comment by pedroigor** [comment]
Looks like you already tried using the old administration console. What I'm wondering now is the behavior if you use legacy v19 (using Wildfly). It should help to understand if it is related to the new distribution or something else that changed in the codebase that introduce this possible degradation.

**Comment by mortenjoenby** [comment]
Having "truncated" the EVENT_ENTITY table did not make any difference.
Admin Console doesn't work.

**Comment by mortenjoenby** [comment]
Any guidance on how to troubleshoot this is GREATLY appreciated.
We have two things to try:
1) Using v19 with Wildfly
2) Using a database with 70 realms, that we works fine, and then try adding more an more realms

But there must be a way to troubleshoot what Keycloak is doing?

**Comment by mortenjoenby** [comment]
Browser network debugging.
![KC19_Admin_console_debugging](https://user-images.githubusercontent.com/62574327/186381050-d8c5b580-16bd-4ade-a5ee-09c159a32365.jpg)

**Comment by pedroigor** [comment]
@mortenjoenby A lot of work has been done here https://github.com/keycloak/keycloak/discussions/11074 to identify the gaps when using multiple realms. You can find some references there.

If you are able to, please try Legacy v19 to help us to eliminate the distribution from the possibilities. If the same behavior is there, something changed in the code base that introduces this degradation.

**Comment by ladytaz-rowell** [comment]
I am having this same problem using a brand new empty database with nothing but the master realm. I am following the optimized instructions from https://www.keycloak.org/server/containers. I am also using a MSSQL database. Please help resolve this.

**Comment by ladytaz-rowell** [comment]
Here is a snapshot of the console loading and the logs
![image](https://user-images.githubusercontent.com/54676734/186437132-16e992a8-2eb5-4d00-90c9-86f9cd90c427.png)

**Comment by ladytaz-rowell** [comment]
I was able to get this to finally work by adding this to my Dockerfile. I had removed it accidentally trying to figure out another issue.

ENV KC_HOSTNAME=localhost

**Comment by dbarentine** [comment]
I have the same problem on a brand new database and not running in Docker. Setting the hostname or disabling strict hostname doesn't fix it for me.

Interestingly I only see the problem when running in production mode. If I start the server using start-dev it then works.

**Comment by mortenjoenby** [comment]
Hi @pedroigor .
Thanks a lot for your feedback and also the link to the discussion with improvement suggestions. It's not clear to me how much (if any) of that has already made it into KC19?
I have an update on our work/test activities ...

1) We have not tried Wildfly. We most likely won't. Not quite sure.
2) We have tried with 3 databases with 70, 296 and 333 realms.
2a) Seems that +300 starts to become a problem

Yesterday we were honestly at a loss as suddenly it seemed to work pretty well, but we couldn't really explain why :( We then decided today to step back, and start with the default settings.
So what we have tried is the following:

Database with 333 realms, PostgreSQL (Azure Paas) with 2 vCores.
1) k8s startupProbe failureThreshold: 60 (default) ->120: Upgrade fails due to timeout after 5 minutes. Container dies. We were not sure what timeout we were hitting
1a) Considering server in a "faulty" state since upgrade dies. Trying a second time, the "Update database" log entry is NOT shown.
1b) restoring database ...

2) quarkus.transaction-manager.default-transaction-timeout:  60sec (default) -> PT20M (20 mins)
2a) Upgrade succeeds in first attempt. Not sure where the 5 minute timeout came from? But perhaps it's many small transaction lasting longer than 60 seconds? 
2b) Admin Console takes a long time to load, realm list takes a long time to load - EVERY time we refresh browser (is there no cache "surviving"? Total load time is 1.5-2 minutes!

3) Scaling Node size from 2 to 4 CPU's / 8 to 16GB RAM
3a) No change in realm load time

4) Configuring Resource limits on pod
4a) No change in realm load time

5) Setting -XX:MaxRAMPercentage=80.0 (default 50%)
5a) No change in load time

6) "Emptying" very large EVENT_ENTITY table (used pg_dump to export DB without the table, and recreated the table definition after pg_restore)
6a) No change in load time
6b) Did also try with a DB _including_ the table and then setting "expiration time" on User events to let Keycloak cleanup, which actually works after a while. We see load on Postgres, but the #rows on the table drops dramatically.

7) cache-conf-file: Default cache size is 10000 entries. We are not really sure we hit a limit, but we are not sure if this should be tuned?
7a) Tried setting it to 1000000, but didn't seem to make a difference

One thing we know for sure is a must is to change the timeout on the proxy (in this case Azure Application Gateway) from 30secs to e.g. 300secs. Otherwise Admin Console simply fails when loading of realms takes too long!

Other than this we can see that performance seems much better when using the database with 296 realm instead of 333. Not a big difference one might say, but performance wise it seems to make a difference.

We have tried so many things and we are not 100% what the "right" setup should be!
Since we are migrating a several servers to KC19 (and upgrading to PostgreSQL 14 from 11) and one of them has 333 realms, we might just have to accept the bad load time in Admin Console. Keep in mind we haven't even tested the authentication part from our application yet. That's tomorrow and next week :) Hopefully that works fine, and Admin Console is not too important. But we still use API's to create new realms (not on that instance anymore though).

Not sure if this is of any help, or any guidance can be provided, but please just shoot - any suggestions and ideas are welcome.

Thanks!

**Comment by mortenjoenby** [comment]
Oh, one more thing. We see these "errors" (INFO level) in the logs. Not clear to us if that's a problem, but logging stacktraces at INFO level ...? Is that normal?
[TransactionReaperWorker_messages.txt](https://github.com/keycloak/keycloak/files/9425941/TransactionReaperWorker_messages.txt)

**Comment by mortenjoenby** [comment]
@pedroigor, one more thing (second actually :)) ...
We are quite sure that when I created this issue, and the Admin Console wouldn't load "at all", was 99% certainty the timeout in the proxy (AGW).

**Comment by souravs17031999** [comment]
@mortenjoenby Hi, I was using keycloak version 18 and migrating to 19 and I also saw these thread dumps being printed in the logs at "INFO" level, not sure why is that.
Any help is appreciated.

I will attach the logs here
[thread logs arjuna.txt](https://github.com/keycloak/keycloak/files/9427108/thread.logs.arjuna.txt)

**Comment by pedroigor** [comment]
Regarding the stack traces, I would expect similar entries in logs if you are using `18`. We fixed an issue with the health endpoints recently that should fix this problem. And It was quite easy to reproduce: load the server with requests to `/health`. And the server won't shut down because of being blocked by arjuna.

But this one is not so good:

```
WARN  [com.arjuna.ats.arjuna] (Transaction Reaper Worker 0) ARJUNA012095: Abort of action id 0:ffff646e7f84:85bb:63074fa2:6b9 invoked while multiple threads active within it.
```

I'll try reproducing this by using the `keycloak-benchmark` project.

**Comment by pedroigor** [comment]
@souravs17031999 @mortenjoenby  I was able to reproduce the problem related to the stack trace you provided. Created https://github.com/keycloak/keycloak/issues/14011.

**Comment by mortenjoenby** [comment]
Hi @pedroigor .
Regarding the `quarkus.transaction-manager.default-transaction-timeout`, can you provide any guidance what the value should be? And perhaps also explain why we had to change it?
So default value is 60 seconds, and we changed it to 20mins (PT20M), but 20 mins is just a value we grabbed out of thin air. The upgrade took ~7 mins as described. So should we also change it back to the default 60 seconds once upgrade is successful?

And about the https://github.com/keycloak/keycloak/discussions/11074, when can (some of) these improvements be expected to make it into an official release?

Thanks,
Morten

**Comment by pedroigor** [comment]
The transaction timeout should be ideally a lower value. The default value should be enough. A high value might cause requests to be blocked in case you are facing some temporary degradation and cause cascading failures.

However, during migration, you might have operations taking more time and reaching the timeout limit. So yes, you can have it higher enough to allow migration but lower afterward after the migration happens.

W.r.t. to #11074, based on the discussions it should be expected. But it involves a lot of collaboration, discussions, and different PRs. Not sure when it will happen.

**Comment by mortenjoenby** [comment]
Thanks @pedroigor .
We will change it back to default after upgrade.
I will leave the issue open until next week, but I guess that we haven't got any issues right now.

**Comment by mortenjoenby** [comment]
Hi @pedroigor .
Are there any recommendations to e.g. amount of CPU's / servers (clustered setup) in relation to amount of realms?
It seems that when loading the database with 330+ realms on a DEV setup with one node of 2 CPU's is most of the time an issue. I am for example trying to delete realms via Admin Console, which I can, but it takes quite a long time for the console to "realize" that it's deleted and then refreshing the list of realms.

Also I am wondering how the caching works? One the realms have initially loaded are they then not in a cache? It just seems that when I refresh my browser, it fetches everything again from database (I think), otherwise I don't understand why it takes so long.

Thanks again,
Morten

**Comment by pedroigor** [comment]
There are several reasons why it takes so long. One of them is how the server handles requests from the admin console when it reaches the `whoami` endpoint. It goes over every single realm and performs a lot of operations to resolve access.

FYI, we have a team working on benchmarking and optimizations to the server. If you did not know yet, we have a sub-project called `https://github.com/keycloak/keycloak-benchmark` that can be used to share benchmarking results and report issues.

I propose we close this issue so that you can report your findings there.

**Comment by mortenjoenby** [comment]
Thanks @pedroigor .
To be honest I am not too familiar with the `whoami` endpoint and how/where this is used.
I will look at that `keycloak-benchmark` project. 
As a last note, I can inform you that when I delete realm and the count drops to 290-300, performs starts getting much better.
Thanks.

**Comment by souravs17031999** [comment]
@pedroigor regarding stack traces above, appreciate #14011.
This also means that liveness and readiness probes which goes to /health/live and /health/ready can cause these traces ?
I have deployed in K8S cluster where these probes are getting hit every 15 seconds with timeout of 5 seconds and I beleive this is the reason I am seeing lot of these traces there and even I saw some events generated in the pod which showed me time to time these probes are throwing 500 error status code.

Will this impact KC performance in any way or deployment stability in terms of containers health ?

**Comment by pedroigor** [comment]
@souravs17031999 If you are using v19+ the issue should *only* happen if you set `http-relative-path` other than `/`.

But yes, the issue that was fixed was causing transactions to not complete properly (requests to health/metrics are dispatched to a thread other than the worker thread to which the request was dispatched to) and eventually timeout. A symptom of this issue is blocking when shutting down the server.

**Comment by pedroigor** [comment]
@mortenjoenby I'm wondering if we can close this issue and track the improvements you are requesting in https://github.com/keycloak/keycloak/discussions/11074?

**Comment by mortenjoenby** [comment]
Hi @pedroigor .
Following on @souravs17031999 's comment, is this why we are then seeing our pods restarting due to unhealthy Liveness probes?
Warning  Unhealthy  4m50s (x563 over 22h)  kubelet  Liveness probe failed: HTTP probe failed with statuscode: 500

We are luckily running in a 3 node k8s cluster, but see this happening MANY times :(

**Comment by mortenjoenby** [comment]
@pedroigor , we would really like to know if this is a problem with Keycloak not functioning correctly? Or should we for now "just" change the retry interval and the timeout value (15 and 5)? If that makes a difference at all?
And when can we expect 19.0.2 to be released?

**Comment by souravs17031999** [comment]
@mortenjoenby even with 15 and 5, I am seeing lot of thread stacktraces being printed and time to time 500 status code but pods are not restarting due to failureThreshold of 3 but I don't know for sure how long will it hold true if we start increasing load on Keycloak server, will it start restarting pods.

What I beleive is that it should not hamper Keycloak incoming requests even on high load but @pedroigor can suggest better if it's a real problem and is there some fix we can do

**Comment by mortenjoenby** [comment]
Hi @souravs17031999 , we are now trying with the following changes:

livenessProbe:
  failureThreshold: 3->10
  httpGet:
    path: /auth/health/live
    port: http
    scheme: HTTP
  periodSeconds: 10->30
  successThreshold: 1
  timeoutSeconds: 5->10
  
Liveness probe of course still fails, but it takes more for the container to be restarted.
And then we need v19.0.2 asap :)

**Comment by souravs17031999** [comment]
well, container restarting is definitely a problem :( , but what we have to see is that only requests to /auth/health/live is failing or does it also interferes with other incoming requests like token endpoints.

If it's only health endpoints that are affected, then somehow we can try to find some stable K8S configuration of probes which might depend on how we have setup the keycloak deployment and try to make it not restart , maybe increasing periodSeconds might help.

**Comment by pedroigor** [comment]
@mortenjoenby @souravs17031999 The 500 status code you might be facing when probing the server should be related to https://github.com/keycloak/keycloak/pull/14227.

It is targeted for 19.0.2 and it only happens if you are using a `http-relative-path` other than `/`.

**Comment by souravs17031999** [comment]
@pedroigor Yes, I can confirm atleast I am using ```/auth```
